client/session.go

```diff
@@ -781,24 +834,64 @@ func (s *session) writeAttempt(
 		return errSessionStateNotOpen
 	}
 
+	var (
+		enqueued int32
+		majority = atomic.LoadInt32(&s.majority)
+	)
+
+	// NB(prateek): We retain an individual copy of the namespace, ID per
+	// writeState, as each writeState tracks the lifecycle of it's resources in
+	// use in the various queues. Tracking per writeAttempt isn't sufficient as
+	// we may enqueue multiple writeStates concurrently depending on retries
+	// and consistency level checks.
+	nsID := s.idPool.Clone(namespace)
+	tsID := s.idPool.Clone(id)
+	var tagEncoder serialize.Encoder
+	if wType == taggedWriteAttemptType {
+		tagEncoder = s.tagEncoderPool.Get()
+		if err := tagEncoder.Encode(inputTags); err != nil {
+			tagEncoder.Finalize()
+			s.RUnlock()
+			return err
+		}
+	}
+
 	state := s.writeStatePool.Get()
 	state.topoMap = s.topoMap
 	state.incRef()
 
-	op := s.writeOperationPool.Get()
-	op.namespace = nsID
-	op.request.ID = tsID.Data().Get()
-	op.shardID = s.topoMap.ShardSet().Lookup(tsID)
-	op.request.ID = tsID.Data().Get()
-	op.request.Datapoint.Value = value
-	op.request.Datapoint.Timestamp = timestamp
-	op.request.Datapoint.TimestampTimeType = timeType
-	op.request.Datapoint.Annotation = annotation
-	op.completionFn = state.completionFn
-
-	// todo@bl: Can we combine the writeOperationPool and the writeStatePool?
+	var op writeOp
+	switch wType {
+	case untaggedWriteAttemptType:
+		wop := s.writeOperationPool.Get()
+		wop.namespace = nsID
+		wop.shardID = s.topoMap.ShardSet().Lookup(tsID)
+		wop.request.ID = tsID.Data().Get()
+		wop.request.Datapoint.Value = value
+		wop.request.Datapoint.Timestamp = timestamp
+		wop.request.Datapoint.TimestampTimeType = timeType
+		wop.request.Datapoint.Annotation = annotation
+		op = wop
+	case taggedWriteAttemptType:
+		wop := s.writeTaggedOperationPool.Get()
+		wop.namespace = nsID
+		wop.shardID = s.topoMap.ShardSet().Lookup(tsID)
+		wop.request.ID = tsID.Data().Get()
+		wop.request.EncodedTags = tagEncoder.Data().Get()
+		wop.request.Datapoint.Value = value
+		wop.request.Datapoint.Timestamp = timestamp
+		wop.request.Datapoint.TimestampTimeType = timeType
+		wop.request.Datapoint.Annotation = annotation
+		op = wop
+	default:
+		// should never happen
+		return errUnknownWriteAttemptType
+	}
+
+	// todo@bl: Can we combine the writeOpPool and the writeStatePool?
 	state.op, state.majority = op, majority
-	state.nsID, state.tsID = nsID, tsID
+	state.nsID, state.tsID, state.tagEncoder = nsID, tsID, tagEncoder
+	op.SetCompletionFn(state.completionFn)
```

Comment:
Hm, this is going to allocate as it will need to take a ref to the state ptr as well as the fn ptr.

Reply:
I don't follow what you mean by allocate. 

I ran the escape analysis compiler dump using `go build -gcflags '-m -l'`  on both the `op.SetCompletionFn(state.completionFn)` (the new way) and `op.completionFn = state.completionFn` (the old way); they both mark that `state.completionFn` escapes. 

Is there another allocation you're concerned about?

Reply:
This is fine, I think there is a small alloc here but looking at an `alloc_space` flame graph it looks like it's miniscule.